pkg/apis/clusteroperator/types.go

```diff
@@ -348,6 +359,9 @@ type MachineSetConfig struct {
 type MachineSetSpec struct {
 	// MachineSetConfig is the configuration for the MachineSet
 	MachineSetConfig
+
+	// Version references the clusterversion the machine set is running.
+	Version ClusterVersionReference
```

Comment:
As I think more about this, this needs to be an `ObjectReference` that points to an actual distinct existing `ClusterVersion` object. `ClusterStatus` should also have an `ObjectReference` that points to the same `ClusterVersion` object. The way it is now, if a `ClusterVersion` object is deleted and a new one is created with the same name, then a `Cluster` and its `MachineSets` can be out of sync with each other.

Reply:
Should we keep ClusterVersionReference and add UID, or revert to ObjectReference? I'm assuming we keep and add UID.

When you mention ClusterStatus having a ref to same ClusterVersion, did you mean ClusterSpec? Or do you suggest ClusterVersionReference on ClusterSpec (without UID set), require that to resolve to a UID and get set again on ClusterStatus, which then unblocks machine set creation. What does that layer in the middle buy vs just requiring ClusterSpec.ClusterVerRef set with UID?

Reply:
My suggestion is this.
- `ClusterSpec` has a `ClusterVersionRef` field that is a `ClusterVersionReference`, which includes a namespace and a name. This is set by the user to indicate which `ClusterVersion` object the user wants.
- `ClusterStatus` has a `ClusterVersionRef` field that is an `ObjectReference`, which points to a specific existing `ClusterVersion` object with the name and namespace specified in the `ClusterVersionRef` from the `ClusterSpec`. This is set by the ClusterController when the `ClusterVersion` object matching the requested namespace and name is found. This is cleared by ClusterController when the `ClusterVersion` object is deleted.
- `MachineSetSpec` has a `ClsuterVersionRef` field that is an `ObjectReference`, which points to a specific existing `ClusterVersion` object. This is set by the ClusterController when the MachineSet is created.

> What does that layer in the middle buy vs just requiring ClusterSpec.ClusterVerRef set with UID?
We cannot require `ClusterSpec.ClusterVersionRef` to be set with a UID. The user *must* be able to create the Cluster object prior to the ClusterVersion object existing. That eventual reconciliation is a core tenet of kubernetes. And a controller should not modify the spec of objects that it does not create.

Reply:
I am also fine with the `ClusterVersionRef` in `ClusterStatus` being just a `UID`, as that is sufficient to uniquely identify the ClusterVersion object. The ClusterVersion object can be found using the `ClusterVersionRef` in `ClusterSpec`.

Reply:
Ok thanks this makes more sense to me now.

Reply:
I would also be fine with the `ClusterVersionRef` in `MachineSetSpec` being an object that only includes the name, namespace, and UID, since the other fields of `ObjectReference` are unnecessary for our purposes. I don't know what the conventional way to do this is. We may need to get some outside guidance.

Reply:
I'm a bit worried about the fact that this appears to imply that whenever a cluster version is created, we have to iterate every cluster looking to see if it tried to reference that version and update it's status if necessary. I expect a CI deployment would be a lot of version creations but few clusters, and a real world deployment a lot of clusters but few creations of versions, so maybe this is ok.

Reply:
Yes, it was always going to be the case where we would have to iterate over all of a type of resource when a cluster version was created. With the original implementation, we would have had to iterate over all of the machine sets when a cluster version was created. Otherwise there would be no way for the machine set controller to know when it was OK to sync a machine set if the cluster version were created after the machine set.